analysis/deep2_r23_o32/plotting/plot_difference_curvefit.py

```diff
@@ -0,0 +1,284 @@
+    secondorder_polynomial, thirdorder_polynomial, threevariable_fit, \
+    bian18_R23_OH, bian18_O32_OH, bian18_OH_O32, jiang_O32_OH_fit, jiang18_coeffs
+
+
+def plot_difference_threevariable(lR23, lO32, OH, lO32_all, pdf_file,
+                                  fitting_model, bin_start, bin_end,
+                                  new_coefficients=[], n_bins=4, data_err=[],
+                                  OH_err=[], OH_range=[7,11],
+                                  data_range=[-0.3, 0.3],
+                                  marker=[], label=[], IDs=[], log=None):
+    """
+    So this curve fitting function looks like it fits for three variables fits
+    a * x**3 + b*x**2 + c * x + d*lO32
+    This is also going to be used for the Jiang calibration
+
+
+    Plot differences between LACR23 vs observed R23 as
+    a function of metallicity
+    Plot differences between LACO32 vs observed O32 as
+    a function of metallicity
+    Used by main() function
+    """
+    # print('new coeff: ', new_coefficients, type(new_coefficients))
+    if log is None:
+        log = log_stdout()
+
+    # log.info("starting ...")
+
+    fig, ax = plt.subplots()
+
+    n_sample = len(lR23)
+
+    # Plotting
+    if len(marker) == 0:
+        marker = ['o'] * n_sample
+
+    diff0 = []
+    for nn in range(n_sample):
+        if fitting_model == 'jiang':
+            fitted_function = jiang_O32_OH_fit((OH[nn], lO32[nn]),
+                                               *jiang18_coeffs)
+            log.info(f"jiang_R23: {fitted_function}")
+        if fitting_model == 'Zcal':
+            if len(new_coefficients) != 0:
+                fitted_function = threevariable_fit((OH[nn], lO32[nn]),
+                                                    *new_coefficients)
+                # log.info(f"curve fit LAC_R23: {fitted_function}")
+
+        if nn == 0:
+            if IDs:
+                for jj in range(len(fitted_function)):
+                    id_diff = lR23[nn][jj] - fitted_function[jj]
+                    ax.annotate(IDs[nn][jj], (OH[nn][jj], id_diff),
+                                fontsize='6')
+
+        # Label in upper left the points
+        if len(label) != 0:
+            x1 = OH_range[0] + 0.025 * (OH_range[1] - OH_range[0])
+            y1 = data_range[1] - (nn * 0.035 + 0.05) \
+                     * (data_range[1] - data_range[0])
+            x2 = OH_range[0] + 0.035 * (OH_range[1] - OH_range[0])
+            y2 = data_range[1] - (nn * 0.035 + 0.0525) \
+                     * (data_range[1] - data_range[0])
+            ax.text(x2, y2, label[nn], fontsize=8, va='center', ha='left')
+            ax.plot([x1], [y1], marker=marker[nn], color='black')
+
+        for ii in range(n_bins):
+            y_ii_min = bin_start[ii]
+            y_ii_max = bin_end[ii]
+            idx = np.where((lO32[nn] >= y_ii_min) &
+                           (lO32[nn] <= y_ii_max))[0]
+
+            ii_label = ''
+            if nn == 2:  # n_sample-1:
```

Comment:
I would avoid referring to specific indices. It will break if the order of input changes. Use a naming convention for each dataset instead. Like if `name == 'DEEP2'`

Reply:
fixed